modules/gapi/src/executor/gasync.cpp

```diff
@@ -0,0 +1,184 @@
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2019 Intel Corporation
+
+#include "opencv2/gapi/gcomputation_async.hpp"
+#include "opencv2/gapi/gcomputation.hpp"
+#include "opencv2/gapi/gcompiled_async.hpp"
+#include "opencv2/gapi/gcompiled.hpp"
+
+#include <condition_variable>
+
+#include <future>
+#include <condition_variable>
+//#include <chrono>
+#include <stdexcept>
+#include <queue>
+
+namespace {
+    //This is a tool to move initialize captures of a lambda in C++11
+    template<typename T>
+    struct move_through_copy{
+       T value;
+       move_through_copy(T&& g) : value(std::move(g)) {}
+       move_through_copy(move_through_copy&&) = default;
+       move_through_copy(move_through_copy const& lhs) : move_through_copy(std::move(const_cast<move_through_copy&>(lhs))) {}
+    };
+}
+
+namespace cv {
+namespace gapi {
+namespace wip  {
+
+    namespace impl{
+        class async_service {
+
+            std::mutex mtx;
+            std::condition_variable cv;
+            std::queue<std::function<void()>> q;
+            std::atomic<bool> exiting;
+
+            std::thread thrd;
+
+        public:
+            async_service() = default ;
+
+            void add_task(std::function<void()>&& t){
+                if (!thrd.joinable())
+                {
+                    //thread has not been started yet, so start it
+                    thrd = std::thread {[this](){
+                        decltype(q) second_q;
+                        while (!exiting){
+                            std::unique_lock<std::mutex> lck{mtx};
+                            if (q.empty())
+                            {
+                                cv.wait(lck, [&](){ return exiting || !q.empty();});
+                            }
+
+                            q.swap(second_q);
+                            lck.unlock();
+
+                            if (!second_q.empty())
+                            {
+                                auto& f = second_q.front();
+                                f();
+                                second_q.pop();
+                            }
+                        }
+                    }};
+
+                }
+                std::unique_lock<std::mutex> lck{mtx};
+                bool first_task = q.empty();
+                q.push(std::move(t));
+                lck.unlock();
+                if (first_task)
+                {
+                    cv.notify_one();
+                }
+            }
+            ~async_service(){
+                if (thrd.joinable())
+                {
+                    exiting = true;
+                    mtx.lock();
+                    mtx.unlock();
+                    cv.notify_one();
+                    thrd.join();
+                }
+            }
+        };
+
+        async_service the_ctx;
+    }
```

Comment:
This whole block of code still lacks comments. "What author has wanted to say?"

Reply:
see below at [line 128](https://github.com/opencv/opencv/pull/14346/files/93213d9b36121c61cec7b4a1297bd437125d5f00#diff-a6eac27ed59a18bd5f4bebfe2a80efe5R128)

Is it enough - or more comments are required ?

Reply:
fixed the code and added comments

Reply:
I still don't understand what is happening here. Queue, second queue (local instance). What is the purpose of this async_service? Why it's so complicated?

Reply:
Is there any reply on @alalek's concerns regarding https://github.com/opencv/opencv/pull/14346#discussion_r276182722 ?